cstrml/market/src/lib.rs

```diff
@@ -429,16 +457,66 @@ impl<T: Trait> Module<T> {
                     T::Payment::pay_sorder(order_id);
                 }
 
-                // TODO: add market slashing here
-                // TODO: Record `failed_count` from `Success` to `Failed`
-                // TODO: Record `failed_duration` from `Failed` to `Success`
-
                 // 2. Update storage order
                 <StorageOrders<T>>::insert(order_id, so);
             }
         }
     }
 
+    pub fn maybe_punish_provider(order_id: &<T as system::Trait>::Hash) {
+        // 1. Update record's count
+        let so = Self::storage_orders(order_id).unwrap();
+        let mut punishment = Self::provider_punishments(order_id).unwrap();
+        let punish_duration = T::PunishDuration::get();
+        if so.status == OrderStatus::Success {
+            punishment.success += 1;
+        } else if so.status == OrderStatus::Failed {
+            punishment.failed += 1;
+        } else {
+
+        }
+        if punishment.success + punishment.failed >= punish_duration && punishment.punished_value < so.amount {
+            // 2. do slash
+            let real_punish_value = Self::do_slash_provider(&so, &punishment, &punish_duration);
+            punishment.punished_value += real_punish_value;
+            // TODO: add closed status here.
+            // 3. reset
+            punishment.success = 0;
+            punishment.failed = 0;
+        } else {
+
+        }
+        <ProviderPunishments<T>>::insert(&order_id, punishment);
+    }
+
+    pub fn do_slash_provider(
```

Comment:
using slash here since it's a slash from balance perspective

Reply:
No, I think we should unify all ***Code Concept***. Just named `punish_provider` is okay.

Reply:
No, I think we should unify all ***Code Concept***. 
Moreover, this function did lots of punishment dealing logic.
As my perspective, just named `punish_provider` is fine.